library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
+    def upload(self):
+        software_path = self.want.software
+        hotfix_path = self.want.hotfix
+        image_list = self.list_images_on_device()
+        hotfix_list = self.list_hotfixes_on_device()
 
-    def software_active(self, activity):
-        result = False
-        images = self.api.System.SoftwareManagement.get_all_software_status()
+        if self.want.forced:
+            self.remove()
 
-        hotfix = self.params['hotfix']
-        software = self.params['software']
+        if software_path and not self.image_exists_on_device():
+            if self.want.remote_src:
+                self.download_iso_on_device()
+            else:
+                self.upload_to_device(software_path)
+            self.wait_for_images(image_list)
 
-        if hotfix:
-            photfix = self.params['photfix']
+        if hotfix_path and not self.hotfix_exists_on_device():
+            if self.want.remote_src:
+                self.download_iso_on_device(True)
+            else:
+                self.upload_to_device(hotfix_path)
+            self.wait_for_images(hotfix_list, True)
 
-        if software:
-            psoftware = self.params['psoftware']
+    def remove(self):
+        software_path = self.want.software
+        hotfix_path = self.want.hotfix
+        image_list = self.list_images_on_device()
+        hotfix_list = self.list_hotfixes_on_device()
 
-        for image in images:
-            ibuild = image['base_build']
-            iver = image['version']
-            iprod = image['product']
+        if hotfix_path:
+            self.delete_hotfix_on_device()
+            self.wait_for_images(hotfix_list, True)
 
-            if image['active'] == activity:
-                if hotfix:
-                    pbuild = photfix['build']
-                    pver = photfix['version']
-                    pprod = photfix['product']
+        if software_path:
+            self.delete_image_on_device()
+            self.wait_for_images(image_list)
 
-                    if pbuild == ibuild and pver == iver and pprod == iprod:
-                        result = True
+    def exists(self):
+        forced = self.want.forced
+        software_path = self.want.software
+        hotfix_path = self.want.hotfix
 
-                if software:
-                    pbuild = psoftware['build']
-                    pver = psoftware['version']
-                    pprod = psoftware['product']
+        if forced:
+            return False
 
-                    if pbuild == ibuild and pver == iver and pprod == iprod:
-                        result = True
+        if software_path and hotfix_path:
+            if self.image_exists_on_device() and \
+             self.hotfix_exists_on_device():
+                return True
+        elif hotfix_path:
+            if self.hotfix_exists_on_device():
+                return True
+        elif software_path:
+            if self.image_exists_on_device():
+                return True
```

Comment:
you dont have a false to the true tests in those blocks.

```
        if software_path and hotfix_path:
            if self.image_exists_on_device() and self.hotfix_exists_on_device():
                return True
```

where is the case handled where if the first line is true, but the second line is false. The code would fall through and return None instead of False.